pytential/linalg/skeletonization.py

```diff
@@ -0,0 +1,790 @@
+    """Builds a block matrix that can be used to skeletonize the
+    rows (targets) or columns (sources) of the symbolic matrix block
+    described by ``(ibrow, ibcol)``.
+    """
+
+    if indices.nblocks == 1:
+        raise ValueError("cannot make a proxy skeleton for a single block")
+
+    # {{{ generate proxies
+
+    domain = wrangler.domains[ibcol]
+    discr = places.get_discretization(domain.geometry, domain.discr_stage)
+    pxy = proxy_generator(actx, domain, indices)
+
+    # }}}
+
+    # {{{ evaluate farfield (proxy) and nearfield (neighbor) interactions
+
+    from pytential.linalg import gather_block_neighbor_points
+    nbrindices = gather_block_neighbor_points(
+            actx, discr, pxy,
+            max_particles_in_box=max_particles_in_box)
+
+    pxymat, pxyindices = evaluate_farfield(
+            actx, places, pxy, nbrindices, ibrow=ibrow, ibcol=ibcol)
+    nbrmat, nbrindices = evaluate_nearfield(
+            actx, places, pxy, nbrindices, ibrow=ibrow, ibcol=ibcol)
+
+    # }}}
+
+    # {{{ concatenate everything to get the blocks ready for ID-ing
+
+    def block_take(idx, mat, i):
+        istart, iend = idx._block_ranges[i:i + 2]
+        return mat[istart:iend].reshape(idx.block_shape(i, i))
+
+    pxyblk = np.full((indices.nblocks, indices.nblocks), 0, dtype=object)
+    for i in range(indices.nblocks):
+        pxymat_i = block_take(pxyindices, pxymat, i)
+        nbrmat_i = block_take(nbrindices, nbrmat, i)
+
+        if pxyindices.row is indices:
+            # NOTE: targets [P, N]
+            pxyblk[i, i] = np.hstack([pxymat_i, nbrmat_i])
+        else:
+            # NOTE: sources [P; N]
+            pxyblk[i, i] = np.vstack([pxymat_i, nbrmat_i])
+
+    # }}}
+
+    return pxyblk
+
+
+def _skeletonize_block_by_proxy_with_mats(
+        actx, ibcol, ibrow, places, proxy_generator, wrangler, blkindices,
+        id_eps=None, id_rank=None,
+        max_particles_in_box=None):
+    L = np.full((blkindices.nblocks, blkindices.nblocks), 0, dtype=object)
+    R = np.full((blkindices.nblocks, blkindices.nblocks), 0, dtype=object)
+
+    if blkindices.nblocks == 1:
+        L[0, 0] = np.eye(blkindices.row.indices.size)
+        R[0, 0] = np.eye(blkindices.col.indices.size)
+
+        return L, R, blkindices
+
+    # construct proxy matrices to skeletonize
+    src_mat = _make_block_proxy_skeleton(
+            actx, ibrow, ibcol,
+            places, proxy_generator, wrangler, blkindices.col,
+            evaluate_farfield=wrangler.evaluate_source_farfield,
+            evaluate_nearfield=wrangler.evaluate_source_nearfield,
+            max_particles_in_box=max_particles_in_box)
+    tgt_mat = _make_block_proxy_skeleton(
+            actx, ibrow, ibcol,
+            places, proxy_generator, wrangler, blkindices.row,
+            evaluate_farfield=wrangler.evaluate_target_farfield,
+            evaluate_nearfield=wrangler.evaluate_target_nearfield,
+            max_particles_in_box=max_particles_in_box)
```

Comment:
Factor out the commonalities using `partial`?

Reply:
Done in 40815f6.